x/erc20/keeper/denom_metadata_hooks.go

```diff
@@ -0,0 +1,39 @@
+package keeper
+
+import (
+	"fmt"
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+)
+
+type ERC20BankContractRegistrationHook struct {
+	erc20Keeper Keeper
+}
+
+// NewERC20ContractRegistrationHook returns the DenomMetadataHooks for ERC20 registration
+func NewERC20ContractRegistrationHook(erc20Keeper Keeper) ERC20BankContractRegistrationHook {
+	return ERC20BankContractRegistrationHook{
+		erc20Keeper: erc20Keeper,
+	}
+}
+
+func (e ERC20BankContractRegistrationHook) AfterDenomMetadataCreation(ctx sdk.Context, newDenomMetadata banktypes.Metadata) error {
```

Comment:
I think it would be good to explain the whole feature in a docstring on this function

To make sure I understand, is the following text correct?

```
We assume that whoever is sending IBC packets with a token denom X, already did a CreateDenomMetadata for that token. At that time, the hook will be fired and the contract will have been created.

Therefore, when the packet arrives, the contract is already created, and we can convert to erc20 freely
```

I still don't understand why we have the positive balance check.

Also, is something needed for the reverse direction, erc20->ibc?

Reply:
> We assume that whoever is sending IBC packets with a token denom X, already did a CreateDenomMetadata for that token.

I don't think it has to be the same user. In order to create the denom, you need to have a special permission. To do an IBC transfer, you don't need it.

But the idea is more less that, first register the denom in the bank, and in the erc20 module (with the contract), and only after that you can successfully make IBC transfers for that denom. If the denom (metadata) is missing in the bank, or the contract is not created for whatever reason, the IBC transfer should fail.

The positive balance check is just a courtesy check I thought might be nice, so we don't go any further when there's no balance on the user's account. I'm not 100% sure we need it though

Reply:
I do agree it should be documented. 

As for the erc20-> ibc, that's a good question. For @omritoptix. Because I have no idea

Reply:
once erc20 contract exists, the conversion is automatic.
to create ERC20 contract for IBC tokens, this PR handles it
to turn a new ERC20 to IBC, a `RegisterERC20` should be called, which is currently triggered by gov only.